workflow/scripts/rule_site_density_mask.py

```
@@ -0,0 +1,109 @@
+        # to add the first site position to get the absolute position
+        first_site = all_positions[0]
+        start += first_site
+        end += first_site
+
+        # Find the index of start and end positions
+        start = numpy.argmax(all_positions >= start)
+        end = numpy.argmax(all_positions >= end)
+        site_density_mask[:start] = True
+        site_density_mask[end:] = True
+    site_density_mask_key = make_filter_key(
+        snakemake.wildcards.subset_name,
```

Ruff (0.14.5)

Undefined name `snakemake`

(F821)